// jtg/views/turtle-graphics.jsx
/*
 * Copyright (c) 2015 by Greg Reimer <gregreimer@gmail.com>
 * All Rights Reserved.
 */

var Turtle = require('jtg/models/turtle')
  , React = require('react')

/*
Environment({
  width: 'auto',
  height: 'auto',
  bgColor: '#000',
  fgColor: '#000',
  commandsPerFrame: 1
}, (t) => {})
*/

module.exports = React.createClass({

  componentDidMount() {

    var field = this.refs.field.getDOMNode()
      , draw = this.refs.draw.getDOMNode()
      , turtle = this.refs.turtle.getDOMNode()

    var t = window.t = new Turtle({
      fieldCanvas: field,
      drawCanvas: draw,
      turtleCanvas: turtle
    })

    Turtle.run(function *() {
      for (var i=0; i < 1500000; i++) {
        for (var j=0; j < 360; j++) {
          yield t.fd(i / 90)
          yield t[i%2?'lt':'rt'](1)
        }
      }
    })
  },

  render() {

    var ratio = window.devicePixelRatio
      , width = this.props.width
      , height = this.props.height
      , style = { width: width + 'px', height: height + 'px' }

    return <div className="turtle-graphics" style={style}>
      <canvas width={width * ratio} height={height * ratio} ref="field"/>
      <canvas width={width * ratio} height={height * ratio} ref="draw"/>
      <canvas width={width * ratio} height={height * ratio} ref="turtle"/>
    </div>
  }
})
