// jtg/models/pen.js
/*
 * Copyright (c) 2015 by Greg Reimer <gregreimer@gmail.com>
 * All Rights Reserved.
 */

module.exports = class {

  constructor(turtle) {
    turtle._copyPropsTo(this)
    this._state = {
      isDown: true,
      color: '#fff',
      weight: 1
    }
  }

  isUp() {
    var state = this._state
    return !state.isDown
  }

  isDown() {
    var state = this._state
    return state.isDown
  }

  up() {
    var seq = this._sequencer
      , state = this._state
    seq.runNoDelay(function(){
      state.isDown = false
    }, this)
  }

  down() {
    var seq = this._sequencer
      , state = this._state
    seq.runNoDelay(function(){
      state.isDown = true
    }, this)
  }

  weight(len) {
    var state = this._state
    if (arguments.length === 0) {
      return state.weight
    } else {
      let seq = this._sequencer
      return seq.runNoDelay(function() {
        state.weight = len
      }, this)
    }
  }

  color(color) {
    var state = this._state
    if (arguments.length === 0) {
      return state.color
    } else {
      let seq = this._sequencer
      return seq.runNoDelay(function() {
        state.color = color
      }, this)
    }
  }
}
