// jtg/models/pen.js
/*
 * Copyright (c) 2015 by Greg Reimer <gregreimer@gmail.com>
 * All Rights Reserved.
 */

module.exports = class {

  constructor(turtle) {
    turtle._copyPropsTo(this)
    this._state = {
      isDown: true,
      color: '#fff',
      weight: 1
    }
  }

  isUp() {
    var state = this._state
    return !state.isDown
  }

  isDown() {
    var state = this._state
    return state.isDown
  }

  up() {
    var state = this._state
    return () => state.isDown = false
  }

  down() {
    var state = this._state
    return () => state.isDown = true
  }

  weight(len) {
    var state = this._state
    if (arguments.length === 0) {
      return state.weight
    } else {
      return () => state.weight = len
    }
  }

  flood() {
    return () => {
      var ctx = this._drawContext
        , grid = this._grid
      ctx.fillStyle = this.color()
      ctx.fillRect(0, 0, grid.width, grid.height)
    }
  }

  color(color) {
    var state = this._state
    if (arguments.length === 0) {
      return state.color
    } else {
      return () => state.color = color
    }
  }
}
